python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -270,3 +271,346 @@ def to_numpy(self, zero_copy_only: bool = True):
+    def from_numpy(
+        cls, arr: Union[np.ndarray, List[np.ndarray], Tuple[np.ndarray]]
+    ) -> "ArrowRaggedTensorArray":
+        """
+        Convert an ndarray or an iterable of heterogeneous-shaped ndarrays to an array
+        of heterogeneous-shaped, homogeneous-typed tensors.
+
+        Args:
+            arr: An ndarray or an iterable of heterogeneous-shaped ndarrays.
+
+        Returns:
+            An ArrowRaggedTensorArray containing len(arr) tensors of heterogeneous
+            shape.
+        """
+        # Implementation note - Arrow representation of ragged tensors:
+        #
+        # We represent an array of ragged tensors using a struct array containing two
+        # fields:
+        #  - data: a variable-sized list array, where each element in the array is a
+        #    tensor element stored in a 1D (raveled) variable-sized list of the
+        #    underlying scalar data type.
+        #  - shape: a variable-sized list array containing the shapes of each tensor
+        #    element.
+        if isinstance(arr, Iterable):
+            if isinstance(arr, np.ndarray) and arr.dtype.type is not np.object_:
+                raise ValueError(
+                    "ArrowRaggedTensorArray should only be used for "
+                    "heterogeneous-shaped tensor elements, i.e. of object dtype, but "
+                    f"got dtype: {arr.dtype}"
+                )
+            arr = list(arr)
+        elif not isinstance(arr, (list, tuple)):
+            raise ValueError(
+                "ArrowRaggedTensorArray can only be constructed from an ndarray or a "
+                f"list/tuple of ndarrays, but got: {type(arr)}"
+            )
+        if len(arr) == 0:
+            # Empty ragged tensor arrays are not supported.
+            raise ValueError("Creating empty ragged tensor arrays is not supported.")
+
+        # Whether all subndarrays are contiguous views of the same ndarray.
+        dtypes, shapes, sizes, raveled = [], [], [], []
+        for a in arr:
+            a = np.asarray(a)
+            dtypes.append(a.dtype)
+            shapes.append(a.shape)
+            sizes.append(a.size)
+            # Convert to 1D array view; this should be zero-copy in the common case.
+            # NOTE: If array is not in C-contiguous order, this will convert it to
+            # C-contiguous order, incurring a copy.
+            a = np.ravel(a, order="C")
+            raveled.append(a)
+        dtype = np.find_common_type([], dtypes)
+        if dtype is None:
+            raise ValueError(f"Unable to find a common dtype for dtypes: {dtypes}")
+        pa_dtype = pa.from_numpy_dtype(dtype)
+        # Get size offsets and total size.
+        sizes = np.array(sizes)
+        size_offsets = np.cumsum(sizes)
+        total_size = size_offsets[-1]
+        # Concatenate 1D views into a contiguous 1D array.
+        if (
+            all(_is_contiguous_view(curr, prev) for prev, curr in _pairwise(raveled))
```

Comment:
if we're using `all` here for pairwise element, which i think is already O(n) operation, why not further safe guard this to ensure our `_is_ragged_tensor` function also checks shape of all element rather than just optimistically O(1) by checking the first element ?

From previous segmentation model CUJ i found it's quite common in larger dataset with blocks / table concat that the first element looks fine but last / middle element actually have different shape / type.

Reply:
A few things here:
1. This O(n) pairwise check is only done once, on construction, and we're already doing other linear passes here. The `_is_ragged_tensor()` check currently happens whenever the dtype is requested, hence the hope to keep it as an O(1) check. We could always cache the check on construction since `TensorArray` should be immutable.
2. The best-effort `_is_ragged_tensor()` check is only checking if the ndarray is of object dtype and its elements are ndarrays; it's not checking to see if all elements are of different shape. Since object-dtyped tensors can sneak in, this check probably needs to be more rigorous in general, e.g. checking that (1) the ndarray is of object dtype, (2) the elements are all ndarrays, (3) the elements all have the same dtype.

Reply:
Agreed on the context of 1) , for 2) looks like there's no O(1) way to know all elements dtype based on your description. What if we adopt caching the check as metadata to the block if the immutability contract holds ? So upon seeing TensorArray we always know if it's ragged or not in subsequent processing.